### PR TITLE
make_flux_sparsity_pattern with flux couplings on a subset of the faces

### PR DESCRIPTION
**Pull request by simonsticko**
Add another make_flux_sparsity_pattern function which takes an
additional std::function describing over which faces there should be
flux couplings in the created sparsity pattern.

### CODE REVIEW

File: source/dofs/dof_tools_sparsity.cc [hunk 1]
```diff
@@ -759,20 +759,24 @@ namespace DoFTools
     {
       // implementation of the same function in namespace DoFTools for
       // non-hp DoFHandlers
-      template <typename DoFHandlerType,
+      template <int dim,
+                int spacedim,
```
Comment (bangerth) [inline source/dofs/dof_tools_sparsity.cc]: I have to admit that I don't quite understand what this change is about -- aren't you implementing a function here that is unrelated to the one you declare in the .h file, given that they have different template arguments and function argument types?

Comment (simonsticko) [reply source/dofs/dof_tools_sparsity.cc]: After the changes in the first commit of this pull-req the MSVC-build failed. The reason was that in the hp-case MVSC couldn't deduce whether it should call the function taking the generic DoFHandlerType or the specialization for the hp::DoFHandler. Honestly, I don't understand why. It compiled both with gcc and clang without warnings. Adding the changes in the second commit was the only way I could come up with that would definitely make the MSVC-build pass.

Comment (bangerth) [reply source/dofs/dof_tools_sparsity.cc]: But it doesn't do what you want :-)

Think about how that file looks like once you just `#include` the .h file into the .cc file:
```
// from the .h file
namespace DoFTools {
  template <typename DoFHandlerType>
  void some_function (const DoFHandlerType &dh);
}

// now what you have in the .cc file:
namespace DoFTools {
  template <int dim>
  void some_function (const DoFHandler<dim> &dh)
  {
    ... your implementation...
  }
}
```
In essence, you end up with two overloaded functions that happen to share the same name. If a user calls the first of these two, you will find that there is implementation of that function to be found; conversely, a user can't call the second function because it has never been declared in a header file.

So maybe MSVC compiles this, but it doesn't do what you hope.

**Comment by bangerth** [comment]
Just remove the new function again and instead add the new argument to the existing function. That should avoid the trouble you see.

File: include/deal.II/dofs/dof_tools.h [hunk 2]
```diff
@@ -775,18 +775,41 @@ namespace DoFTools
...
+    const AffineConstraints<number> &constraints,
+    const bool                       keep_constrained_dofs,
+    const Table<2, Coupling> &       couplings,
+    const Table<2, Coupling> &       face_couplings,
+    const types::subdomain_id        subdomain_id,
+    const std::function<
+      bool(const typename DoFHandlerType::active_cell_iterator &,
+           const unsigned int)> &face_has_flux_coupling =
+      [](const typename DoFHandlerType::active_cell_iterator &,
+         const unsigned int) { return true; });
```
Comment (bangerth) [inline include/deal.II/dofs/dof_tools.h]: Yes, this change looks good.

File: source/dofs/dof_tools_sparsity.cc [hunk 3]
```diff
@@ -763,13 +763,17 @@ namespace DoFTools
...
+        const DoFHandlerType &           dof,
+        SparsityPatternType &            sparsity,
+        const AffineConstraints<number> &constraints,
+        const bool                       keep_constrained_dofs,
+        const Table<2, Coupling> &       int_mask,
+        const Table<2, Coupling> &       flux_mask,
+        const types::subdomain_id        subdomain_id,
+        const std::function<
+          bool(const typename DoFHandlerType::active_cell_iterator &,
+               const unsigned int)> &face_has_flux_coupling)
```
Comment (bangerth) [inline source/dofs/dof_tools_sparsity.cc]: And so do the changes to this function in the `internal` namespace.

File: source/dofs/dof_tools_sparsity.cc [hunk 4]
```diff
@@ -1069,7 +1076,11 @@ namespace DoFTools
         const bool                                   keep_constrained_dofs,
         const Table<2, Coupling> &                   int_mask,
         const Table<2, Coupling> &                   flux_mask,
-        const types::subdomain_id                    subdomain_id)
+        const types::subdomain_id                    subdomain_id,
+        const std::function<
+          bool(const typename dealii::hp::DoFHandler<dim, spacedim>::
+                 active_cell_iterator &,
+               const unsigned int)> &face_has_flux_coupling)
```
Comment (bangerth) [inline source/dofs/dof_tools_sparsity.cc]: and this one for the hp case in the `internal` namespace

File: source/dofs/dof_tools_sparsity.cc [hunk 5]
```diff
@@ -1345,17 +1359,23 @@ namespace DoFTools
...
+    const DoFHandlerType &           dof,
+    SparsityPatternType &            sparsity,
+    const AffineConstraints<number> &constraints,
+    const bool                       keep_constrained_dofs,
+    const Table<2, Coupling> &       int_mask,
+    const Table<2, Coupling> &       flux_mask,
+    const types::subdomain_id        subdomain_id,
+    const std::function<
+      bool(const typename DoFHandlerType::active_cell_iterator &,
+           const unsigned int)> &face_has_flux_coupling)
```
Comment (bangerth) [inline source/dofs/dof_tools_sparsity.cc]: and then this one is again the global function you want to modify, so this looks alright to me as well.

File: source/dofs/dof_tools_sparsity.cc [hunk 6]
```diff
@@ -1390,16 +1410,22 @@ namespace DoFTools
              "associated DoF handler objects, asking for any subdomain other "
              "than the locally owned one does not make sense."));
 
+    Assert(
+      face_has_flux_coupling,
+      ExcMessage(
+        "The function which specifies if a flux coupling occurs over a given "
+        "face is empty."));
```
Comment (bangerth) [inline source/dofs/dof_tools_sparsity.cc]: nice job adding the assertion!

**Comment by bangerth** [comment]
The patch looks good to me, subject of course to the error message you see. I suspect that that really just reflects a compiler bug. The symbol names you see can be "demangled" using `c++filt` and look like this:
```
_ZNK6dealii8DoFToolsUlRKNS_18TriaActiveIteratorINS_15DoFCellAccessorINS_10DoFHandlerILi1ELi1EEELb0EEEEEjE_clES8_j
dealii::DoFTools::{lambda(dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::DoFHandler<1, 1>, false> > const&, unsigned int)#1}::operator()(dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::DoFHandler<1, 1>, false> > const&, unsigned int) const
_ZNK6dealii8DoFToolsUlRKNS_18TriaActiveIteratorINS_15DoFCellAccessorINS_2hp10DoFHandlerILi1ELi1EEELb0EEEEEjE_clES9_j
dealii::DoFTools::{lambda(dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::hp::DoFHandler<1, 1>, false> > const&, unsigned int)#1}::operator()(dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::hp::DoFHandler<1, 1>, false> > const&, unsigned int) const
_ZNK6dealii8DoFToolsUlRKNS_18TriaActiveIteratorINS_15DoFCellAccessorINS_10DoFHandlerILi2ELi2EEELb0EEEEEjE_clES8_j
dealii::DoFTools::{lambda(dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::DoFHandler<2, 2>, false> > const&, unsigned int)#1}::operator()(dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::DoFHandler<2, 2>, false> > const&, unsigned int) const
_ZNK6dealii8DoFToolsUlRKNS_18TriaActiveIteratorINS_15DoFCellAccessorINS_2hp10DoFHandlerILi2ELi2EEELb0EEEEEjE_clES9_j
dealii::DoFTools::{lambda(dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::hp::DoFHandler<2, 2>, false> > const&, unsigned int)#1}::operator()(dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::hp::DoFHandler<2, 2>, false> > const&, unsigned int) const
```
So I think that what is happening is that the compiler creates the same lambda function every time the file is compiled, and then complains that the same lambda function is defined multiple times. This makes no sense, but seems to be what is happening.

You can work around this by putting something like this at the top of the file:
```
namespace DoFTools
{
  namespace internal
  {
     template <typename DoFHandlerType>
     inline
     bool always_couple_on_faces (...your argument list...)
     {
       return true;
     }
  }
}
```
and then instead of using the lambda function as default argument, just do `= &internal::always_couple_on_faces<DoFHandlerType>`.

Does that help?

**Comment by simonsticko** [comment]
> Does that help?

Yes, that worked out nicely. :+1: I have added it.

However, after adding the new parameter, MVSC can't deduce which of the two functions in `DoFTools::internal::` that should be called:
```
[00:07:43] C:\projects\source\source\dofs\dof_tools_sparsity.cc(1419,15): error C2668: 'dealii::DoFTools::internal::`anonymous-namespace'::make_flux_sparsity_pattern': ambiguous call to overloaded function [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:07:43] C:\projects\source\source\dofs\dof_tools_sparsity.cc(1072,7): message : could be 'void dealii::DoFTools::internal::`anonymous-namespace'::make_flux_sparsity_pattern<1,1,SparsityPatternType,double>(const dealii::hp::DoFHandler<1,1> &,SparsityPatternType &,const dealii::AffineConstraints<double> &,const bool,const dealii::Table<2,dealii::DoFTools::Coupling> &,const dealii::Table<2,dealii::DoFTools::Coupling> &,const dealii::types::subdomain_id,const std::function<bool (const dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::hp::DoFHandler<1,1>,false>> &,unsigned int)> &)' [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:07:43]           with
[00:07:43]           [
[00:07:43]               SparsityPatternType=dealii::SparsityPattern
[00:07:43]           ]
[00:07:43] C:\projects\source\source\dofs\dof_tools_sparsity.cc(766,7): message : or       'void dealii::DoFTools::internal::`anonymous-namespace'::make_flux_sparsity_pattern<DoFHandlerType,SparsityPatternType,double>(const DoFHandlerType &,SparsityPatternType &,const dealii::AffineConstraints<double> &,const bool,const dealii::Table<2,dealii::DoFTools::Coupling> &,const dealii::Table<2,dealii::DoFTools::Coupling> &,const dealii::types::subdomain_id,const std::function<bool (const dealii::TriaActiveIterator<dealii::DoFCellAccessor<DoFHandlerType,false>> &,unsigned int)> &)' [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:07:43]   Generating fe_q_bubbles.inst
[00:07:43]           with
[00:07:43]           [
[00:07:43]               DoFHandlerType=dealii::hp::DoFHandler<1,1>,
[00:07:43]               SparsityPatternType=dealii::SparsityPattern
[00:07:43]           ]
[00:07:43] C:\projects\source\source\dofs\dof_tools_sparsity.cc(1426,1): message : while trying to match the argument list '(const DoFHandlerType, SparsityPatternType, const dealii::AffineConstraints<double>, const bool, const dealii::Table<2,dealii::DoFTools::Coupling>, const dealii::Table<2,dealii::DoFTools::Coupling>, const dealii::types::subdomain_id, const std::function<bool (const dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::hp::DoFHandler<1,1>,false>> &,unsigned int)>)' [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:07:43]           with
[00:07:43]           [
[00:07:43]               DoFHandlerType=dealii::hp::DoFHandler<1,1>
[00:07:43]           ]
[00:07:43]           and
[00:07:43]           [
[00:07:43]               SparsityPatternType=dealii::SparsityPattern
[00:07:43]           ]
[00:07:43] C:\projects\source\build\source\dofs\dof_tools_sparsity.inst(3252): message : see reference to function template instantiation 'void dealii::DoFTools::make_flux_sparsity_pattern<dealii::hp::DoFHandler<1,1>,dealii::SparsityPattern,double>(const dealii::hp::DoFHandler<1,1> &,dealii::SparsityPattern &,const dealii::AffineConstraints<double> &,const bool,const dealii::Table<2,dealii::DoFTools::Coupling> &,const dealii::Table<2,dealii::DoFTools::Coupling> &,const dealii::types::subdomain_id,const std::function<bool (const dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::hp::DoFHandler<1,1>,false>> &,unsigned int)> &)' being compiled [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
```
[MSVC -appveyor-log.txt](https://github.com/dealii/dealii/files/4102240/MSVC.-appveyor-log.txt)

It's unclear to me why this is more difficult for MSVC than before adding the parameter. :confused:

**Comment by bangerth** [comment]
That's now a bug in MSVC :-( But I think it should be relatively easy to work around. The function in line 766 of the .cc file is unnecessarily general: It takes a `DoFHandlerType` as argument, but because there is a specialization for `hp::DoFHandler` in line 1072, the only possible value for `DoFHandlerType` is `dealii::DoFHandler<dim,spacedim>`.

Try to adjust the template argument list of the function in line 766 to look the same as the one in 1072, with the only difference of the type of the first argument of the two. I imagine it should then work.

**Comment by simonsticko** [comment]
> Try to adjust the template argument list of the function in line 766 to look the same as the one in 1072, with the only difference of the type of the first argument of the two. I imagine it should then work.

Yes, I thought that would have worked to. :confused: This fix worked when I overloaded `make_flux_sparsity_pattern` (in the original pull-req). But when using a default lambda for the new parameter, it seems that MSVC can't handle this either: 
[MSVC -appveyor-log.txt](https://github.com/dealii/dealii/files/4108289/MSVC.-appveyor-log.txt)
```
[00:06:12] C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\functional(1004,1): error C2514: 'std::_Func_impl_no_alloc<decay<_Fx>::type,_Ret,const dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::DoFHandler<1,1>,false>>&,unsigned int>': class template cannot be constructed [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:06:12]           with
[00:06:12]   Generating fe_values.impl.1.inst
[00:06:12]           [
[00:06:12]               _Fx=remove_reference<bool(__cdecl *&)(const DoFHandlerType::active_cell_iterator &,unsigned int)>::type,
[00:06:12]               _Ret=bool
[00:06:12]           ]
[00:06:12] C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\functional(1009): message : see declaration of 'std::_Func_impl_no_alloc<decay<_Fx>::type,_Ret,const dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::DoFHandler<1,1>,false>>&,unsigned int>' [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:06:12]           with
[00:06:12]           [
[00:06:12]               _Fx=remove_reference<bool(__cdecl *&)(const DoFHandlerType::active_cell_iterator &,unsigned int)>::type,
[00:06:12]               _Ret=bool
[00:06:12]           ]
[00:06:12] C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\functional(1130): message : see reference to function template instantiation 'void std::_Func_class<_Ret,const dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::DoFHandler<1,1>,false>> &,unsigned int>::_Reset<remove_reference<bool(__cdecl *&)(const DoFHandlerType::active_cell_iterator &,unsigned int)>::type>(_Fx &&)' being compiled [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:06:12]           with
[00:06:12]           [
[00:06:12]               _Ret=bool,
[00:06:12]               _Fx=remove_reference<bool(__cdecl *&)(const DoFHandlerType::active_cell_iterator &,unsigned int)>::type
[00:06:12]           ]
[00:06:12] C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\functional(1130): message : see reference to function template instantiation 'void std::_Func_class<_Ret,const dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::DoFHandler<1,1>,false>> &,unsigned int>::_Reset<remove_reference<bool(__cdecl *&)(const DoFHandlerType::active_cell_iterator &,unsigned int)>::type>(_Fx &&)' being compiled [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:06:12]           with
[00:06:12]           [
[00:06:12]               _Ret=bool,
[00:06:12]               _Fx=remove_reference<bool(__cdecl *&)(const DoFHandlerType::active_cell_iterator &,unsigned int)>::type
[00:06:12]           ]
[00:06:12] C:\projects\source\include\deal.II/dofs/dof_tools.h(832): message : see reference to function template instantiation 'std::function<bool (const dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::DoFHandler<1,1>,false>> &,unsigned int)>::function<bool(__cdecl *)(const DoFHandlerType::active_cell_iterator &,unsigned int),enable_if<std::conjunction_v<std::negation<std::is_same<decay<bool(__cdecl *&)(const DoFHandlerType::active_cell_iterator &,unsigned int)>::type,std::function<bool (const dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::DoFHandler<1,1>,false>> &,unsigned int)>>>,std::_Is_invocable_r<_Ret,bool(__cdecl *&)(const DoFHandlerType::active_cell_iterator &,unsigned int),const dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::DoFHandler<1,1>,false>>&,unsigned int>>,void>::type>(_Fx)' being compiled [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:06:12]           with
[00:06:12]           [
[00:06:12]               _Ret=bool,
[00:06:12]               _Fx=bool (__cdecl *)(const DoFHandlerType::active_cell_iterator &,unsigned int)
[00:06:12]           ]
[00:06:12] C:\projects\source\source\dofs\dof_tools_sparsity.cc(1350): message : see reference to function template instantiation 'std::function<bool (const dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::DoFHandler<1,1>,false>> &,unsigned int)>::function<bool(__cdecl *)(const DoFHandlerType::active_cell_iterator &,unsigned int),enable_if<std::conjunction_v<std::negation<std::is_same<decay<bool(__cdecl *&)(const DoFHandlerType::active_cell_iterator &,unsigned int)>::type,std::function<bool (const dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::DoFHandler<1,1>,false>> &,unsigned int)>>>,std::_Is_invocable_r<_Ret,bool(__cdecl *&)(const DoFHandlerType::active_cell_iterator &,unsigned int),const dealii::TriaActiveIterator<dealii::DoFCellAccessor<dealii::DoFHandler<1,1>,false>>&,unsigned int>>,void>::type>(_Fx)' being compiled [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:06:12]           with
[00:06:12]           [
[00:06:12]               _Ret=bool,
[00:06:12]               _Fx=bool (__cdecl *)(const DoFHandlerType::active_cell_iterator &,unsigned int)
[00:06:12]           ]
[00:06:12] C:\projects\source\build\source\dofs\dof_tools_sparsity.inst(10416): message : see reference to function template instantiation 'void dealii::DoFTools::make_flux_sparsity_pattern<dealii::DoFHandler<1,1>,dealii::SparsityPattern>(const dealii::DoFHandler<1,1> &,dealii::SparsityPattern &,const dealii::Table<2,dealii::DoFTools::Coupling> &,const dealii::Table<2,dealii::DoFTools::Coupling> &,const dealii::types::subdomain_id)' being compiled [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:06:12] C:\projects\source\bundled\boost-1.70.0\include\boost/bind/placeholders.hpp(54): message : see reference to class template instantiation 'boost::arg<9>' being compiled [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:06:12] C:\projects\source\bundled\boost-1.70.0\include\boost/bind/placeholders.hpp(53): message : see reference to class template instantiation 'boost::arg<8>' being compiled [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:06:12] C:\projects\source\bundled\boost-1.70.0\include\boost/bind/placeholders.hpp(52): message : see reference to class template instantiation 'boost::arg<7>' being compiled [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:06:12] C:\projects\source\bundled\boost-1.70.0\include\boost/bind/placeholders.hpp(51): message : see reference to class template instantiation 'boost::arg<6>' being compiled [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:06:12] C:\projects\source\bundled\boost-1.70.0\include\boost/bind/placeholders.hpp(50): message : see reference to class template instantiation 'boost::arg<5>' being compiled [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:06:12] C:\projects\source\bundled\boost-1.70.0\include\boost/bind/placeholders.hpp(49): message : see reference to class template instantiation 'boost::arg<4>' being compiled [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:06:12] C:\projects\source\bundled\boost-1.70.0\include\boost/bind/placeholders.hpp(48): message : see reference to class template instantiation 'boost::arg<3>' being compiled [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:06:12] C:\projects\source\bundled\boost-1.70.0\include\boost/bind/placeholders.hpp(47): message : see reference to class template instantiation 'boost::arg<2>' being compiled [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
[00:06:12] C:\projects\source\bundled\boost-1.70.0\include\boost/bind/placeholders.hpp(46): message : see reference to class template instantiation 'boost::arg<1>' being compiled [C:\projects\source\build\source\dofs\obj_dofs_debug.vcxproj]
```

File: source/dofs/dof_tools_sparsity.cc [hunk 7]
```diff
@@ -1390,16 +1407,22 @@ namespace DoFTools
...
+
     internal::make_flux_sparsity_pattern(dof,
```
Comment (bangerth) [inline source/dofs/dof_tools_sparsity.cc]: To work around the MSVC issue, try to specify the template arguments explicitly here:
```suggestion
    internal::make_flux_sparsity_pattern<dim,spacedim>(dof,
```
Let's see whether that helps!

Comment (simonsticko) [reply source/dofs/dof_tools_sparsity.cc]: I tried this but it didn't help.

File: source/dofs/dof_tools_sparsity.cc [hunk 8]
```diff
@@ -1347,13 +1347,15 @@ namespace DoFTools
...
+    make_flux_sparsity_pattern(
+      dof,
+      sparsity,
+      dummy,
+      keep_constrained_dofs,
+      int_mask,
+      flux_mask,
+      subdomain_id,
+      internal::always_couple_on_faces<DoFHandlerType>);
   }
```
Comment (simonsticko) [inline source/dofs/dof_tools_sparsity.cc]: I could make the MSVC build pass by calling with the default value here.

**Comment by bangerth** [comment]
It looks like you made it work? Is this ready to merge?

**Comment by bangerth** [comment]
If so, please also squash the three commits and add a changelog entry. Thanks!

**Comment by simonsticko** [comment]
> It looks like you made it work? Is this ready to merge?

Yes

> If so, please also squash the three commits and add a changelog entry. Thanks!

Done!

File: doc/news/changes/minor/20200128SimonSticko [hunk 9]
```diff
@@ -0,0 +1,3 @@
+Add an optional std::function parameter to make_flux_sparsity_pattern which can
+be used to specify over which faces there should be a flux coupling in the
+created sparsity pattern.
```
Comment (bangerth) [inline doc/news/changes/minor/20200128SimonSticko]: Please add your name to the end of the file, as we do in all of the other files. Also don't forget to end the file with a newline!

Comment (simonsticko) [reply doc/news/changes/minor/20200128SimonSticko]: Okey, done.

**Comment by bangerth** [comment]
Thanks for figuring out the details of making MSVC happy!
